### PR TITLE
MDEV-40174: Remove unnecessary double parsing JSON document

### PR DESCRIPTION
**Pull request by VasuBhakt**
Fixes [MDEV-40174](https://jira.mariadb.org/browse/MDEV-40174) 

# Problem
The `json_normalize_engine` function, used by `json_normalize`, calls `json_valid_engine` before processing. The processing reparses the same JSON document that `json_valid` already processed.

# Solution
Removed the redundant `json_valid_engine` pre-check to eliminate double-parsing, and integrated error handling directly into the normalization engine:

1. **Catch Empty Strings:** Return an error directly in `json_normalize_engine` if the root type is `JSON_VALUE_UNINITIALIZED`.

2. **Catch Trailing Garbage on Scalars:** Updated `json_norm_build` to enforce a full scan to the end of the document for scalar values (like numbers/strings) using `while (json_scan_next(je) == 0)`. This ensures that trailing garbage correctly triggers a syntax error, matching the strict validation previously handled by `json_valid_engine`.

3. **Propagate Syntax Errors:** Updated `json_normalize_engine` to explicitly check and fail if `je->s.error` is set during the build phase.

# Tests
Added tests to `mysql-test/main/json_normalize.test` for invalid documents such as empty strings, whitespace, and trailing garbage on scalars just to be sure the new edge cases are fully covered.

# Version
The change is directed at versions `10.11`, `11.4`, `11.8`, and should also fix `12.3`. The PR is based on branch `10.11`, being the lowest affected branch.

### CODE REVIEW

File: strings/json_normalize.c [hunk 1]
```diff
@@ -431,7 +431,6 @@ json_normalize_sort(struct json_norm_value *val)
     break;
   }
   case JSON_VALUE_UNINITIALIZED:
-    DBUG_ASSERT(0);
```
Comment (gkodinov) [inline strings/json_normalize.c:434]: Why do you need to remove this? I don't think you should be removing this.

File: strings/json_normalize.c [hunk 2]
```diff
@@ -579,7 +578,6 @@ json_norm_to_string(DYNAMIC_STRING *buf, struct json_norm_value *val)
   }
   case JSON_VALUE_UNINITIALIZED:
   {
-    DBUG_ASSERT(0);
```
Comment (gkodinov) [inline strings/json_normalize.c:582]: ditto: keep the assert

**Review by gkodinov — CHANGES_REQUESTED** [review]
Thank you for your contribution! This is a preliminary review.

I do not think this is a bug. It's a performance optimization. And, as such, it needs to go to the main branch IMHO. But, this hinges on it being considered not a bug. And this is a determination that the final reviewer should make. 

However, I'm positive that you need to add a commit message that follows contribution guidelines to your commit. So please do. I also have a question below.

File: strings/json_normalize.c (hunk 1 above)
Comment (VasuBhakt) [reply strings/json_normalize.c:434]: You are completely right @gkodinov! Because of the `JSON_VALUE_UNINITIALIZED` check earlier, its impossible for `json_normalize_sort` or `json_norm_to_string` to ever receive an uninitialized value anyway. I overlooked this during my patch and removed those to ensure the server does not crash. However I understand its important that this stays for the sake of defensive programming. I have restored the changes and pushed it here. Let me know what you think!

File: strings/json_normalize.c (hunk 2 above)
Comment (VasuBhakt) [reply strings/json_normalize.c:582]: restored it.

**Comment by VasuBhakt** [comment]
Hi @gkodinov, thanks for the review! You are right about the assert statements, I have reverted them correctly as you said. 
I have also added the commit message as is required by the contribution guidelines.
Regarding the branch to which this change should go, as the Jira ticket mentioned this as a bug with `10.11` being the lowest affected, I based this on `10.11` branch. If @grooverdan wants it to move to `main`, I am good to re-open a new PR  regarding that. Let me know!

File: strings/json_normalize.c [hunk 3]
```diff
@@ -859,6 +857,10 @@ json_norm_build(json_engine_t *je, struct json_norm_value *root,
     err= json_norm_parse(root, je);
     if (err)
       return err;
+  }
```
Comment (grooverdan) [inline strings/json_normalize.c]: remove trailing space on these lines.

File: strings/json_normalize.c [hunk 4]
```diff
@@ -908,14 +912,9 @@ json_normalize_engine(json_engine_t *je, DYNAMIC_STRING *result,
...
 
-
-  if (!json_valid_engine(je, in, in_size, &my_charset_utf8mb4_bin))
-  {
-    err= 1;
-    goto json_normalize_end;
-  }
-
   err= json_norm_build(je, &root, in, in_size, &my_charset_utf8mb4_bin);
+  if (root.type == JSON_VALUE_UNINITIALIZED || je->s.error)
```
Comment (grooverdan) [inline strings/json_normalize.c:916]: What is the cause of JSON_VALUE_UNINITIALIZED? Is there a test case that covers it? Does there need to be an` je->s.error` need to be set?

**Review by grooverdan — COMMENTED** [review]
might need:

```
diff --git a/strings/json_normalize.c b/strings/json_normalize.c
index c72f2c01ccb..dbaf9104cde 100644
--- a/strings/json_normalize.c
+++ b/strings/json_normalize.c
@@ -746,7 +746,7 @@ json_norm_parse(struct json_norm_value *root, json_engine_t *je)
 
 json_norm_parse_end:
   dynstr_free(&key);
-  return err;
+  return err || je->s.err;
 }
```

Can you check that the tests in mysql-test/main/json_normalize.test covers a range of invalid json documents and that errors are generated correctly.

Can you show a performance result for before/after this change?

```
set @obj=concat_ws('','{', repeat('"a":"b",', 1250000/2), '"c":"d"}');
select benchmark(10, JSON_NORMALIZE(@obj));
```

Good update in commit message.

As there's no functional change this isn't a feature and therefore no mandatory requirement for a main target. Poor implementation is notionally a task, as least in my mind, that if it has easy risk free fix is still can be lowest maintained branch, so still 10.11 from my POV.

**Review by gkodinov — APPROVED** [review]
LGTM. Please stand by for the final review.

File: strings/json_normalize.c (hunk 3 above)
Comment (VasuBhakt) [reply strings/json_normalize.c]: fixed it.

File: strings/json_normalize.c (hunk 4 above)
Comment (VasuBhakt) [reply strings/json_normalize.c:916]: Yes the `JSON_VALUE_UNINITIALIZED` is used for catching empty strings from being passed. I added a test in `main.json_normalize` for such a case.
I don't think we need to set `je->s.error` since we handled this manually with `UNINITIALIZED`.

**Comment by VasuBhakt** [comment]
Hi @grooverdan, thanks for the review!

1. I have updated `json_norm_parse_end` to return `err || je->s.error;` as you requested. I also removed the trailing spaces you pointed out in the other comment.
2. I ran the benchmark you provided locally, the results confirm that removing the redundant `json_valid_engine` pre-check actively sped up the parsing slightly (or at worst, caused zero regression):
    * Before patch : ~9.04 s
    * After patch: ~8.75 s
3. I added some tests to `mysql-test/main/json_normalize.test` for invalid documents such as empty strings, whitespace, and trailing garbage on scalars just to be 100% sure the new edge cases are fully covered.

P.S. It looks like `innodb.log_file_overwrite` failed on the CI run, I don't think its because of my code changes though. Kindly let me know if you feel its a problem.

**Review by grooverdan — APPROVED** [review]

**Comment by VasuBhakt** [comment]
Thanks for being such awesome mentors on this one @grooverdan and @gkodinov! Looking forward to making more contributions!
